### PR TITLE
[pred-deadalloc-elim] Teach the pass how to eliminate dead allocation…

### PR DESCRIPTION
**Pull request by gottesmm**
…s that are load [take], store [init] back into memory.

I am doing this to eliminate some differences in codegen before/after
serialization ownership. It just means less of the tests need to be touched when
I flip the switch.

Specifically, today this change allows us to handle certain cases where there is
a dead allocation being used to pass around a value at +1 by performing a load
[take] and then storing a value back into the memory. The general format is an
allocation that only has stores, load [take], and destroy_addr users. Consider
the following SIL:

```
store %x to [init] %mem            (0)
%xhat = load [take] %mem           (1)
%xhat_cast = apply %f(%xhat)       (2)
store %xhat_cast to [init] %mem    (3)
destroy_addr %mem
```

Notice how assuming that we can get rid of the store, we can perform the
following store -> load forwarding:

```
%xhat_cast = apply %f(%x)          (2)
store %xhat_cast to [init] %mem    (3)
destroy_addr %mem
```

In contrast, notice how we get an ownership violation (double consume of %x by
(0) and (2)) if we can not get rid of the store:

```
store %x to [init] %mem
%xhat_cast = apply %f(%x)
store %xhat_cast to [init] %mem    (2)
destroy_addr %mem
```

This is in fact the same condition for promoting a destroy_addr since when a
destroy_addr is a load [take] + destroy_value. So I was able to generalize the
code for destroy_addr to handle this case.

### CODE REVIEW

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Comment by gottesmm** [comment]
@swift-ci smoke test

File: lib/SILOptimizer/Mandatory/PredictableMemOpt.cpp [hunk 1]
```diff
@@ -1551,29 +1553,53 @@ void AllocOptimize::promoteDestroyAddr(
   dai->eraseFromParent();
 }
 
+void AllocOptimize::promoteLoadTake(
+    LoadInst *li, MutableArrayRef<AvailableValue> availableValues) {
+  SILValue address = li->getOperand();
```
Comment (atrick) [inline lib/SILOptimizer/Mandatory/PredictableMemOpt.cpp]: It would be nice to assert that the load is a take.

**Review by atrick — APPROVED** [review]
Looks great. See inline comments.

File: test/SILOptimizer/predictable_deadalloc_elim_ownership.sil [hunk 2]
```diff
@@ -415,3 +418,75 @@ bb0(%0 : @owned $NativeObjectPair, %1 : @owned $Builtin.NativeObject):
...
+  %0hat = load [take] %1 : $*Builtin.NativeObject
+  store %0hat to [init] %1 : $*Builtin.NativeObject
+  destroy_addr %1 : $*Builtin.NativeObject
+  dealloc_stack %1 : $*Builtin.NativeObject
+  %9999 = tuple()
+  return %9999 : $()
+}
+
+// CHECK-LABEL: sil [ossa] @diamond_test_4_with_load_take : $@convention(thin) (@owned Builtin.NativeObject, @owned NativeObjectPair) -> () {
+// CHECK-NOT: alloc_stackfofofofofo
```
[the remark this answers is not in the document]
Comment (gottesmm) [reply test/SILOptimizer/predictable_deadalloc_elim_ownership.sil]: oops.

File: test/SILOptimizer/predictable_deadalloc_elim_ownership.sil [hunk 3]
```diff
@@ -415,3 +418,75 @@ bb0(%0 : @owned $NativeObjectPair, %1 : @owned $Builtin.NativeObject):
...
+  store %0hathat to [init] %1 : $*Builtin.NativeObject
+  destroy_addr %1 : $*Builtin.NativeObject
+  dealloc_stack %1 : $*Builtin.NativeObject
+  %9999 = tuple()
+  return %9999 : $()
+}
+
+// This test makes sure that we first eliminate the destroy_addr, before we
+// eliminate the load [take] since the load [take] value is the available value
+// of the destroy_addr.
```
[the remark this answers is not in the document]
Comment (gottesmm) [reply test/SILOptimizer/predictable_deadalloc_elim_ownership.sil]: It is about the ordering within that single pass. We gather up all of the information and then eliminate the destroy_addr followed by eliminate the load [take].

File: test/SILOptimizer/predictable_deadalloc_elim_ownership.sil [hunk 4]
```diff
@@ -415,3 +418,75 @@ bb0(%0 : @owned $NativeObjectPair, %1 : @owned $Builtin.NativeObject):
   %9999 = tuple()
   return %9999 : $()
 }
+
+// CHECK-LABEL: sil [ossa] @load_take_opt_simple : $@convention(thin) (@owned Builtin.NativeObject) -> () {
+// CHECK: bb0([[ARG:%.*]] : @owned $Builtin.NativeObject):
+// CHECK:   [[RESULT:%.*]] = apply {{%.*}}([[ARG]])
+// CHECK:   destroy_value [[RESULT]]
+// CHECK: } // end sil function 'load_take_opt_simple'
```
[the remark this answers is not in the document]
Comment (gottesmm) [reply test/SILOptimizer/predictable_deadalloc_elim_ownership.sil]: I am just going to add a CHECK-NOT line.

File: lib/SILOptimizer/Mandatory/PredictableMemOpt.cpp [hunk 5]
```diff
@@ -1626,6 +1652,22 @@ bool AllocOptimize::tryToRemoveDeadAllocation() {
       return false;
     case PMOUseKind::InitOrAssign:
       break;    // These don't prevent removal.
+    case PMOUseKind::Load:
+      // For now only handle takes from alloc_stack.
+      //
+      // TODO: It should be implementable, but it has not been needed yet.
```
[the remark this answers is not in the document]
Comment (gottesmm) [reply lib/SILOptimizer/Mandatory/PredictableMemOpt.cpp]: It shouldn't. I just didn't need it and was avoiding writing some more tests. You are right though it is the /right/ thing to do.

**Comment by gottesmm** [comment]
So I looked at the alloc_box issue and turns out what is actually happening is that we do not handle the destroy_value for alloc_box in general. I just didn't remember so I was being conservative. The pass never eliminates alloc_boxes. We could teach it though if we taught it how to understand that a destroy_value of the box is equivalent to a destroy_addr of a stack location.

Since the pass has never supported this, I don't think that is within the scope of this PR.

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Pull request merged**